File: 2021/python/5.py
def count_overlaps(lines, max_x, max_y):
    # start off with nothing in the grid
    grid = [[0 for i in range(max_y + 1)] for j in range(max_x + 1)]

    for start, end in lines:
        # get coordinates
        x_i, y_i = start
        x_f, y_f = end

        # if horizontal line
        if x_i == x_f:
            # work out offset and step size (depending on direction)
            offset = -1 if y_f < y_i else 1
            step = -1 if y_f < y_i else 1

            for y in range(y_i, y_f + offset, step):
                grid[x_i][y] += 1

        # if vertical line
        elif y_i == y_f:
            # work out offset and step size (depending on direction)
            offset = -1 if x_f < x_i else 1
            step = -1 if x_f < x_i else 1
            for x in range(x_i, x_f + offset, step):
                grid[x][y_i] += 1

    # sum up the number of points that have at least 2 overlaps
    overlaps = sum([int(item >= 2) for row in grid for item in row])
    return overlaps


def main():
    lines = []
    max_x, max_y = 0, 0
    with open("../inputs/5.txt", "r") as input:
        for line in input:
            start, end = line.strip().split(" -> ")
            start, end = tuple(map(int, start.split(","))), tuple(map(int, end.split(",")))
            max_x = max(max_x, max(start[0], end[0]))
            max_y = max(max_y, max(start[1], end[1]))
            lines.append([start, end])

    print("PART ONE:", count_overlaps(lines, max_x, max_y))

if __name__ == "__main__":
    main()

File: 2021/julia/5.jl
function count_overlaps(lines, max_x, max_y)
    # start off with nothing in the grid
    grid = [[0 for i in 1:(max_y + 1)] for j in 1:(max_x + 1)]

    for line in lines
        # get coordinates
        start, final = line
        x_i, y_i = start
        x_f, y_f = final

        # slightly wasteful and do everything for x and y
        # work out step size (depending on direction)
        x_step = x_f < x_i ? -1 : 1
        x_range = x_i:x_step:x_f

        y_step = y_f < y_i ? -1 : 1
        y_range = y_i:y_step:y_f

        # if horizontal line
        if x_i == x_f
            for y in y_range
                grid[x_i][y] += 1
            end;

        # if vertical line
        elseif y_i == y_f
            # work out offset and step size (depending on direction)
            for x in x_range
                grid[x][y_i] += 1
            end;

        # if diagonal line
        else
            for xy in zip(x_range, y_range)
                x, y = xy
                grid[x][y] += 1
            end;
        end;
    end;

    # sum up the number of points that have at least 2 overlaps
    overlaps = sum([item >= 2 ? 1 : 0 for row in grid for item in row])
    return overlaps
end;


function main()
    lines = []
    max_x, max_y = 0, 0
    open("../inputs/5.txt", "r") do input
        for line in eachline(input)
            start, final = split(chomp(line), " -> ")
            start, final = tuple(parse.(Int, split(start, ","))...) .+ 1, tuple(parse.(Int, split(final, ","))...) .+ 1
            max_x = max(max_x, max(start[1], final[1]))
            max_y = max(max_y, max(start[2], final[2]))
            push!(lines, [start, final])
        end;
    end;

    println("FULL SOLUTION:", count_overlaps(lines, max_x, max_y))
end;

main()
